### PR TITLE
Use c implementation of secp256k1 in Sign and Verify

### PR DESCRIPTION
**Pull request by auxten**
1. Use c implementation of secp256k1 in Sign and Verify
2. Use LevelDB for metaAckIndexBucket
3. Fix some “DATA RACE”
4. Add crypto/elliptic P224/P256/P384/P512 c-secp256k1 and ed25519 Sign Verify benchmark
5. Add "DATA RACE" detection in logs

### CODE REVIEW

File: crypto/secp256k1/curve.go [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+// See http://www.hyperelliptic.org/EFD/g1p/auto-shortw.html
+type BitCurve struct {
+	P       *big.Int // the order of the underlying field
+	N       *big.Int // the order of the base point
+	B       *big.Int // the constant of the BitCurve equation
+	Gx, Gy  *big.Int // (x,y) of the base point
+	BitSize int      // the size of the underlying field
+}
+
+func (BitCurve *BitCurve) Params() *elliptic.CurveParams {
```
Comment (covenantdb) [inline crypto/secp256k1/curve.go:90]: **[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported method BitCurve.Params should have comment or be unexported

File: crypto/secp256k1/curve.go [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+	y3.Mod(y3, BitCurve.P)
+
+	z3 := new(big.Int).Mul(y, z) //Y1*Z1
+	z3.Mul(big.NewInt(2), z3)    //3*Y1*Z1
+	z3.Mod(z3, BitCurve.P)
+
+	return x3, y3, z3
+}
+
+func (BitCurve *BitCurve) ScalarMult(Bx, By *big.Int, scalar []byte) (*big.Int, *big.Int) {
```
Comment (covenantdb) [inline crypto/secp256k1/curve.go:243]: **[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported method BitCurve.ScalarMult should have comment or be unexported

File: crypto/secp256k1/secp256.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
+func init() {
+	// around 20 ms on a modern CPU.
+	context = C.secp256k1_context_create_sign_verify()
+	C.secp256k1_context_set_illegal_callback(context, C.callbackFunc(C.secp256k1GoPanicIllegal), nil)
+	C.secp256k1_context_set_error_callback(context, C.callbackFunc(C.secp256k1GoPanicError), nil)
+}
+
+var (
+	ErrInvalidMsgLen       = errors.New("invalid message length, need 32 bytes")
```
Comment (covenantdb) [inline crypto/secp256k1/secp256.go]: **[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported var ErrInvalidMsgLen should have comment or be unexported

**Review by xq262144 — APPROVED** [review]
LGTM

File: crypto/secp256k1/secp256.go [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+
+func init() {
+	// around 20 ms on a modern CPU.
+	context = C.secp256k1_context_create_sign_verify()
+	C.secp256k1_context_set_illegal_callback(context, C.callbackFunc(C.secp256k1GoPanicIllegal), nil)
+	C.secp256k1_context_set_error_callback(context, C.callbackFunc(C.secp256k1GoPanicError), nil)
+}
+
+var (
+	// ErrInvalidMsgLen
```
Comment (covenantdb) [inline crypto/secp256k1/secp256.go]: **[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
comment on exported var ErrInvalidMsgLen should be of the form "ErrInvalidMsgLen ..."

File: crypto/secp256k1/secp256.go [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+	ErrInvalidMsgLen = errors.New("invalid message length, need 32 bytes")
+	// ErrInvalidSignatureLen
```
Comment (covenantdb) [inline crypto/secp256k1/secp256.go]: **[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
comment on exported var ErrInvalidSignatureLen should be of the form "ErrInvalidSignatureLen ..."

File: crypto/secp256k1/secp256.go [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+	ErrInvalidSignatureLen = errors.New("invalid signature length")
+	// ErrInvalidRecoveryID
```
Comment (covenantdb) [inline crypto/secp256k1/secp256.go]: **[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
comment on exported var ErrInvalidRecoveryID should be of the form "ErrInvalidRecoveryID ..."

File: crypto/secp256k1/secp256.go [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+	ErrInvalidRecoveryID = errors.New("invalid signature recovery id")
+	// ErrInvalidKey
```
Comment (covenantdb) [inline crypto/secp256k1/secp256.go]: **[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
comment on exported var ErrInvalidKey should be of the form "ErrInvalidKey ..."

File: crypto/secp256k1/secp256.go [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+	ErrInvalidKey = errors.New("invalid private key")
+	// ErrInvalidPubkey
```
Comment (covenantdb) [inline crypto/secp256k1/secp256.go]: **[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
comment on exported var ErrInvalidPubkey should be of the form "ErrInvalidPubkey ..."

File: crypto/secp256k1/secp256.go [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+	ErrInvalidPubkey = errors.New("invalid public key")
+	// ErrSignFailed
```
Comment (covenantdb) [inline crypto/secp256k1/secp256.go]: **[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
comment on exported var ErrSignFailed should be of the form "ErrSignFailed ..."

File: crypto/secp256k1/secp256.go [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+	ErrSignFailed = errors.New("signing failed")
+	// ErrRecoverFailed
```
Comment (covenantdb) [inline crypto/secp256k1/secp256.go]: **[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
comment on exported var ErrRecoverFailed should be of the form "ErrRecoverFailed ..."

**Pull request merged**